### PR TITLE
scheduling and timing

### PR DESCRIPTION
**Pull request by jonwoodruff**
These are a raft of improvements that we have made to our CHERI branch which apply to general performance.

1. Scheduling improvements
Toooba (RISCYOO) previously required that many, many rules in the pipeline conflict with any rule that could report a mispeculation. (Specifically, two ALU rules that finished a misspeculated branch, or some commit rules.) Many rules in the pipeline that consumed speculative state would have to dynamically wait for the conditions of those rules to resolve before they could fire. This both caused timing issues, due to the dynamic dependencies on these rules, and also bluespec scheduling issues, as these rules had to fit into a schedule relative to all the others.  In this patch, the issue is greatly alleviated from multiple angles.  Firstly, the wrongSpec report is deterministically always given priority. (Previously this was left up to the scheduler, and the wrong decision would cost ~10% cycles CoreMark, for example.) This also simplifies the code.  Secondly, the wrongSpec report is buffered in GlobalSpecUpdate so that all rules that are dependent on the broadcast depend only on the one broadcast rule.  Buffering is also added in SpecFifo modules, though changes are applied before any reads in the next cycle, so there is no logical change.  These changes resulted in small amounts of additional improved performance presumably due to decoupling the scheduling of rules that previously were deemed to conflict.

2. Timing improvements.
Several rounds of timing improvements were made in an attempt to reach 50MHz on the Stratix X with two cores. These included buffering and pre-calculation in branch prediction modules, as well as approximating the decode epoch so that prediction could progress more in parallel.

3. Area improvements.
Mainly using NonPipelined versions of the Divider and SquareRooter in the FPU.  This can reduce area by about 10%.  There is also now compression in the BTB, with n ways of compressed targets (where the upper bits (>16) of the target match the branch PC), and a single way is dedicated to full branch targets. This one also improves performance, as an associativity of 2 for the BTB now has an extra way for "far" branches.

These changes have been tested with synthesis (dual core) and booting FreeBSD on the vcu118.

There should be some need to clean up these commits a bit, so feel free to comment!

### CODE REVIEW

File: src_Core/RISCY_OOO/procs/RV64G_OOO/AluExePipeline.bsv [hunk 1]
```diff
@@ -342,6 +343,8 @@ module mkAluExePipeline#(AluExeInput inIfc)(AluExePipeline);
                 mispred: True,
                 isCompressed: x.isCompressed
             });
+            $display("alu mispredict pc¤: %x, nextPc: %x, %d",
```
Comment (jrtc27) [inline src_Core/RISCY_OOO/procs/RV64G_OOO/AluExePipeline.bsv:346]: if (verbose)?

File: src_Core/RISCY_OOO/procs/lib/ReorderBuffer.bsv [hunk 2]
```diff
@@ -595,6 +595,7 @@ module mkSupReorderBuffer#(
                 // move deqP & reset valid
                 deqP[i] <= getNextPtr(deqP[i]);
                 valid[i][deqP[i]][valid_deq_port] <= False;
+                $display("deq[%d][%d]", i, deqP[i]);
```
Comment (jrtc27) [inline src_Core/RISCY_OOO/procs/lib/ReorderBuffer.bsv:598]: if (verbose)?

File: src_Core/RISCY_OOO/procs/lib/ReorderBuffer.bsv [hunk 3]
```diff
@@ -774,6 +775,9 @@ module mkSupReorderBuffer#(
                     end
                 endfunction
                 for(Integer i = 0; i < valueof(SingleScalarSize); i = i+1) begin
+                    if (in_kill_range(fromInteger(i)) != (row[w][i].dependsOn_wrongSpec(specTag) && valid[w][i][valid_wrongSpec_port]))
+                        $display("enqP: %d, enqPNext: %d, w: %d, i: %d, wrongSpec: %d, valid: %d, cur_cycle: %d",
```
Comment (jrtc27) [inline src_Core/RISCY_OOO/procs/lib/ReorderBuffer.bsv:779]: if (verbose)?

File: src_Core/RISCY_OOO/procs/lib/ReorderBuffer.bsv [hunk 4]
```diff
@@ -513,7 +513,7 @@ module mkSupReorderBuffer#(
     Add#(1, a__, aluExeNum), Add#(1, b__, fpuMulDivExeNum)
 );
 
-    Bool verbose = False;
+    Bool verbose = True;
```
Comment (jrtc27) [inline src_Core/RISCY_OOO/procs/lib/ReorderBuffer.bsv:516]: should be left out

**Review by jrtc27 — COMMENTED** [review]
A few related comments after glancing at this

File: src_Core/RISCY_OOO/procs/RV64G_OOO/AluExePipeline.bsv (hunk 1 above)
Comment (jonwoodruff) [reply src_Core/RISCY_OOO/procs/RV64G_OOO/AluExePipeline.bsv]: Done.

**Pull request merged**